### PR TITLE
Process Projects and Tasks

### LINKED ISSUE #745: Implement Tasks response processing
## Description
Process Task response after a successful pull request. Use 3WM extension methods for every sync-able property.
Please note that the `workspace_id` you'll get from pull response will contain `serverId` of the given Workspace not the local one so you will need one extra query to get the workspace's `localId` based on the `serverId` you already have and save the retrieved `localId` with the task. Here is a pseudocode explaining this operation:

```kotlin
val defaultWorkspaceServerId = taskFromPullResponse.default_workspace_id
val defaultWorkspaceLocalId = workspaceDao.getLocalIdByServerId(defaultWorkspaceServerId)
val localTaskWithCorrectRelationToWorkspace = taskFromPullResponse.toLocalTask(defaultWorkspaceLocalId)
```
The same goes for `project_id`!

Xamarin implementation is using one method for processing **tags**, **clients**, **project**, **tasks** and **workspaces** it's called [processEntities](https://github.com/toggl/mobileapp/blob/8779617e49c7da8c9899d36061784dcbb13e4ebf/Toggl.Storage.Realm/Queries/WritePulledData.cs#L220). We probably won't be able to do the same but still have a look for a reference.

## Dependencies
#737  #741

## Definition of done
- [ ] Process sync-able properties using 3WM extension methods
- [ ] Add Tests

### LINKED ISSUE #746: Implement Project response processing
## Description
Process Project response after successful pull request. Use 3WM extension methods for every sync-able property.
Please note that the `workspace_id` you'll get from pull response will contain `serverId` of the given Workspace not the local one so you will need one extra query to get the workspace's `localId` based on the `serverId` you already have and save the retrieved `localId` with the project. Here is a pseudocode explaining this operation:

```kotlin
val defaultWorkspaceServerId = projectFromPullResponse.default_workspace_id
val defaultWorkspaceLocalId = workspaceDao.getLocalIdByServerId(defaultWorkspaceServerId)
val localProjectWithCorrectRelationToWorkspace = projectFromPullResponse.toLocalProject(defaultWorkspaceLocalId)
```
The same goes for `client_id`!

Xamarin implementation is using one method for processing **tags**, **clients**, **project**, **tasks** and **workspaces** it's called [processEntities](https://github.com/toggl/mobileapp/blob/8779617e49c7da8c9899d36061784dcbb13e4ebf/Toggl.Storage.Realm/Queries/WritePulledData.cs#L220). We probably won't be able to do the same but still have a look for a reference.

## Dependencies
#737  #741

## Definition of done
- [ ] Process sync-able properties using 3WM extension methods
- [ ] Add Tests

### PR DESCRIPTION
**Pull request by heytherewill**
### Summary
This adds processing of the pulled projects and tasks to our fetch processing method

### Technical considerations
This is all pretty standard. We needed a specific method for cascade deleting the tasks when deleting the time entries and the projects need some conflict resolution, but other than that everything is trivial

### Relationships
Closes #745
Closes #746 

### Pull Request checklist
- [X] The code does not hinder the app's performance
- [X] The code does not introduce regressions
- [X] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Go at it commit by commit

### Review template
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

## Merge permissions
Anyone, anytime

### CODE REVIEW

File: database/src/main/java/com/toggl/database/dao/SyncDao.kt [hunk 1]
```diff
@@ -14,9 +16,21 @@ interface SyncDao {
...
     @Query("SELECT * FROM clients WHERE serverId = :serverId LIMIT 1")
     suspend fun getClientByServerId(serverId: Long): DatabaseClient?
 
     @Query("SELECT * FROM tags WHERE serverId = :serverId LIMIT 1")
     suspend fun getTagByServerId(serverId: Long): DatabaseTag?
+
+    @Query("SELECT * FROM tasks WHERE serverId = :serverId LIMIT 1")
+    suspend fun getTaskByServerId(serverId: Long): DatabaseTask?
+
+    @Query("SELECT id FROM projects WHERE serverId == :serverId LIMIT 1")
```
Comment (simonrozsival) [inline database/src/main/java/com/toggl/database/dao/SyncDao.kt:31]: `==` in SQL? looks like a typo

File: database/src/main/java/com/toggl/database/dao/SyncDao.kt [hunk 2]
```diff
@@ -14,9 +16,21 @@ interface SyncDao {
     @Query("SELECT * FROM workspaces WHERE serverId == :serverId LIMIT 1")
     suspend fun getWorkspaceByServerId(serverId: Long): DatabaseWorkspace?
 
+    @Query("SELECT id FROM clients WHERE serverId == :serverId LIMIT 1")
```
Comment (simonrozsival) [inline database/src/main/java/com/toggl/database/dao/SyncDao.kt:19]: another `==`

File: repository/src/main/java/com/toggl/repository/Repository.kt [hunk 3]
```diff
@@ -116,7 +116,7 @@ class Repository @Inject constructor(
         )
 
         return projectDao.insert(databaseProject)
-            .run(projectDao::getOne)
+            .let { projectDao.getOne(it) }
```
Comment (simonrozsival) [inline repository/src/main/java/com/toggl/repository/Repository.kt:119]: btw what's the reason for this change?

File: sync/src/main/java/com/toggl/sync/processing/ProcessPulledProjects.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+internal suspend fun processPulledProjects(
+    syncDao: SyncDao,
+    projectDao: ProjectDao,
+    apiProjects: List<ApiProject>
+) {
+    for (apiProject in apiProjects) {
+        val localProject = syncDao.getProjectByServerId(apiProject.id)
+        val isServerDeleted = apiProject.serverDeletedAt != null
+
+        if (localProject == null) {
```
Comment (simonrozsival) [inline sync/src/main/java/com/toggl/sync/processing/ProcessPulledProjects.kt:21]: This `if-else` is getting pretty long, could you extract private `create` and `update` functions from it? Or at least add a `toDatabaseProject` extension similar to `toDatabaseTask`?

File: sync/src/main/java/com/toggl/sync/processing/ProcessPulledProjects.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+                    val updatedWorkspace = localProject.copy(
+                        name = localProject.name updateWith resolvedName,
+                        color = localProject.color updateWith resolvedColor,
+                        active = localProject.active updateWith resolvedActive,
+                        isPrivate = localProject.isPrivate updateWith resolvedIsPrivate,
+                        billable = localProject.billable updateWith resolvedBillable,
+                        workspaceId = localProject.workspaceId updateWith resolvedWorkspaceId,
+                        clientId = localProject.clientId updateWith resolvedClientId,
+                        syncStatus = if (isDirty) EntitySyncStatus.SyncNeeded else EntitySyncStatus.InSync
```
Comment (simonrozsival) [inline sync/src/main/java/com/toggl/sync/processing/ProcessPulledProjects.kt:69]: This is starting to get repeated quite a lot. Could we maybe add a function `syncStatusFromDirtyFlag` or something like that which would contain this if else?

File: sync/src/main/java/com/toggl/sync/processing/ProcessPulledTasks.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    for (apiTask in apiTasks) {
+        val localTask = syncDao.getTaskByServerId(apiTask.id)
+        val isServerDeleted = apiTask.serverDeletedAt != null
+
+        if (localTask == null) {
+            if (!isServerDeleted) {
+                val localWorkspaceId = syncDao.getWorkspaceIdByServerId(apiTask.workspaceId) ?: continue
+                val localProjectId = syncDao.getProjectIdByServerId(apiTask.projectId) ?: continue
+                val newWorkspace = apiTask.toDatabaseTask(localProjectId, localWorkspaceId)
+                taskDao.insert(newWorkspace)
```
Comment (simonrozsival) [inline sync/src/main/java/com/toggl/sync/processing/ProcessPulledTasks.kt:26]:
```suggestion
                val newTask = apiTask.toDatabaseTask(localProjectId, localWorkspaceId)
                taskDao.insert(newTask)
```

**Review by simonrozsival — CHANGES_REQUESTED** [review]

File: database/src/main/java/com/toggl/database/dao/SyncDao.kt (hunk 1 above)
Comment (heytherewill) [reply database/src/main/java/com/toggl/database/dao/SyncDao.kt:31]: 🙈

File: repository/src/main/java/com/toggl/repository/Repository.kt (hunk 3 above)
Comment (heytherewill) [reply repository/src/main/java/com/toggl/repository/Repository.kt:119]: `run` doesn't expect a `suspend fun`

**Review by simonrozsival — APPROVED** [review]
